jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java

```diff
@@ -577,26 +579,20 @@ private static Object[] findJarDirectoryChildren(JarEntryDirectory directory, St
 
     public static IProject getProject(String projectUri) {
         IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
-        IContainer[] containers = root.findContainersForLocationURI(JDTUtils.toURI(projectUri));
-
-        if (containers.length == 0) {
-            return null;
-        }
-
-        // For multi-module scenario, findContainersForLocationURI API may return a container array,
-        // put the result from the nearest project in front.
-        Arrays.sort(containers, (Comparator<IContainer>) (IContainer a, IContainer b) -> {
-            return a.getFullPath().toPortableString().length() - b.getFullPath().toPortableString().length();
-        });
-
-        for (IContainer container : containers) {
-            IProject project = container.getProject();
-            if (!project.exists()) {
-                return null;
+        URI uri = JDTUtils.toURI(projectUri);
+        IContainer[] containers = root.findContainersForLocationURI(uri);
+        
+        Optional<IContainer> maybeProject = Arrays.stream(containers).filter(container -> container instanceof IProject).findFirst();
+        if (maybeProject.isPresent()) {
+        	return (IProject) maybeProject.get();
+        } else {
+            String invisibleProjectName = ProjectUtils.getWorkspaceInvisibleProjectName(FileUtil.toPath(uri).removeTrailingSeparator());
```

Comment:
The current logic here is having some special logic handling the invisible project, but it might not work if there is another kind of project using the linked resources.

I'm thinking that, if the `containers` is not empty but does not have `IProject` instances, can we use `containers[0].getProject()` to the project?

Reply:
I don't understand the use case you are referring to with "if there is another kind of project using the linked resource". 
My current understanding is that there are two scenarios:
- It's either a Gradle or Maven project that has a `IProject` in `containers` (the `if` clause)
- Its an invisible project which is an `IProject` that is linked with a folder to the workspace root and is created by eclipse-jdtls [here](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ProjectUtils.java#L344-L376) in which case `containers` contains the folder that was created by eclipse-jdtls. (the `else` clause)

If there is another scenario that I am currently not aware of, we should add a test case to demonstrate that the code in its current form cannot handle that scenario.  Can you provide a project with the setup that you mentioned or can you point me to a project somewhere on the web that does not work with the current code?

Reply:
Hi @fvclaus,

Sorry for the late reply.

Behind the scenes of the invisible project is actually the linked resource. As you may know, the eclipse-jdtls allows other extension to contribute OSGi bundles as its plugins. Which means, a third-party extension can contribute another project importer, that use the linked resource mechanism as well.

An example could be the jBang extension: https://github.com/jbangdev/jbang-vscode

To test it, you can

1. Install jbang: https://www.jbang.dev/download/
2. install the jbang extension: https://marketplace.visualstudio.com/items?itemName=jbangdev.jbang-vscode
3. In VS Code: F1 > create a new script > (Select a template, for example, the hello world template, and use `.java` as the file extension)
4. Open the folder that contains the generated jbang script file.

After the project is imported, you can see that it uses the linked resourse as well:
<img width="1342" alt="Screenshot 2023-10-03 at 19 28 01" src="https://github.com/microsoft/vscode-java-dependency/assets/6193897/05508855-9d28-40fd-a6e8-a448ac8df5ad">

So, my suggestion is that we can try to avoid assuming that a project will be an invisible project. And use a more generic process to handle it.

Not sure if I declare the problem clearly, let me know if you have more questions :)

Reply:
@jdneo, you are right. Another extension could technically do that.

I checked JBang ([source code](https://github.com/jbangdev/jbang-eclipse/blob/ad7837276d7116ddfa1e7e66a4be5ae609eee8d7/dev.jbang.eclipse.core/src/main/java/dev/jbang/eclipse/core/internal/project/ProjectConfigurationManager.java#L85) and in debugger) and they are creating normal Java Projects inside the project:

![image](https://github.com/microsoft/vscode-java-dependency/assets/2384738/adbffd74-4874-4c9a-a584-f5fd79c4b65a)

I don't understand what the invisible projects are needed for. They code surrounding them is quite complex:
 - Have nature `org.eclipse.jdt.ls.unmanagedFolderNature`
 - Have a location that is not contained in the root paths
 - Have a folder named `_` (WORKSPACE_LINK) that links to the actual location

Otherwise the project will be treated by `java.project.list` as a normal project. But that wouldn't matter anyway, the only thing that would change is the name in the project explorer:
https://github.com/microsoft/vscode-java-dependency/blob/0ec45cc2f67fc31ddefeb7b50680170441d83735/jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ProjectCommand.java#L140-L142

The getProject command can handle URI outside of the workspace root:
https://github.com/microsoft/vscode-java-dependency/blob/0ec45cc2f67fc31ddefeb7b50680170441d83735/jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java#L580

Is it just to keep the compiled artifacts outside of the project?

I changed the code, please have another look.

Reply:
@jdneo, you are referring to the project children. The code I changed only has effect on the top level items in the projects view not on the children. I just checked and it does work, as mentioned in my previous comment, with the current version of the code:
![image](https://github.com/microsoft/vscode-java-dependency/assets/2384738/2bc73eff-39b6-417d-9c12-1a2174a84ae7)

Because the extensions creates a normal Java project, but, as you said, another third party extension might use the invisible project mechanism.

Reply:
Some background of the invisible project:

As you may know, the invisible project is a kind of simple project which does not requires build tools support. Usually, the invisible project can be used for some quick verification, education purpose.

Because we are based on JDT, so the project metadata files(.classpath, .project, .settings/) will be generated at the root of the project. Those files will confuse the student (new users). Then the linked resource mechanism is used for the invisible project. It's just a way to hide those metadata files.